pyodide-build/pyodide_build/build_env.py

```diff
@@ -0,0 +1,230 @@
+    parent_dirs = (Path(curdir) / "garbage").parents[:max_depth]
+
+    for base in parent_dirs:
+        pyproject_file = base / "pyproject.toml"
+
+        if not pyproject_file.is_file():
+            continue
+
+        try:
+            with pyproject_file.open("rb") as f:
+                configs = tomllib.load(f)
+        except tomllib.TOMLDecodeError as e:
+            raise ValueError(f"Could not parse {pyproject_file}.") from e
+
+        if "tool" in configs and "pyodide" in configs["tool"]:
+            return base
+
+    raise FileNotFoundError(
+        "Could not find Pyodide root directory. If you are not in the Pyodide directory, set `PYODIDE_ROOT=<pyodide-root-directory>`."
+    )
+
+
+@functools.cache
+def get_build_environment_vars() -> dict[str, str]:
+    """
+    Get common environment variables for the in-tree and out-of-tree build.
+    """
+    env = _get_make_environment_vars().copy()
+    env["PYODIDE"] = "1"
+
+    if "PYODIDE_JOBS" in os.environ:
+        env["PYODIDE_JOBS"] = os.environ["PYODIDE_JOBS"]
+
+    env["PKG_CONFIG_PATH"] = env["WASM_PKG_CONFIG_PATH"]
+    if "PKG_CONFIG_PATH" in os.environ:
+        env["PKG_CONFIG_PATH"] += f":{os.environ['PKG_CONFIG_PATH']}"
+
+    tools_dir = Path(__file__).parent / "tools"
+
+    env["CMAKE_TOOLCHAIN_FILE"] = str(
+        tools_dir / "cmake/Modules/Platform/Emscripten.cmake"
+    )
+    env["PYO3_CONFIG_FILE"] = str(tools_dir / "pyo3_config.ini")
+
+    hostsitepackages = env["HOSTSITEPACKAGES"]
+    pythonpath = [
+        hostsitepackages,
+    ]
+    env["PYTHONPATH"] = ":".join(pythonpath)
+    return env
+
+
+def get_build_flag(name: str) -> str:
+    """
+    Get a value of a build flag.
+    """
+    build_vars = get_build_environment_vars()
+    if name not in build_vars:
+        raise ValueError(f"Unknown build flag: {name}")
+
+    return build_vars[name]
+
+
+def _set_pyodide_root(*, quiet: bool = False) -> None:
+    """
+    Initialize Pyodide build environment, namely set PYODIDE_ROOT environment variable.
+
+    This function works both in-tree and out-of-tree builds:
+    - In-tree builds: Searches for the root of the Pyodide repository in parent directories
+    - Out-of-tree builds: Downloads and installs the Pyodide build environment into the current directory
+
+    Parameters
+    ----------
+    quiet
+        If True, do not print any messages
+    """
+
+    from . import install_xbuildenv  # avoid circular import
+
+    # If we are building docs, we don't need to know the PYODIDE_ROOT
+    if "sphinx" in sys.modules:
+        os.environ["PYODIDE_ROOT"] = ""
+        return
+
+    # 1) If PYODIDE_ROOT is already set, do nothing
+    if "PYODIDE_ROOT" in os.environ:
+        return
+
+    # 2) If we are doing an in-tree build,
+    #    set PYODIDE_ROOT to the root of the Pyodide repository
+    try:
+        os.environ["PYODIDE_ROOT"] = str(search_pyodide_root(Path.cwd()))
+        return
+    except FileNotFoundError:
+        pass
```

Comment:
This may not be a good idea. If we are doing in tree build, and for some reason that file is not found, it should be a hard failure mentioning that either the wrong command was used by the user  or that it was run from the wrong folder.  We shouldn't try to download `xbuildenv` zip in that case. This means that this command does need to know whether we are doing in tree or out-of-tree build.

Reply:
Yes, probably we should discuss more about how to distinguish in-tree / out-of-tree build. Though,  personally, I think we should go in a direction to merge in-tree / out-of-tree build (which is one of my intention in this PR).

For now, this function is copied from [initialize_pyodide_root](https://github.com/pyodide/pyodide/blob/b80188535396353266ba5f2499450cc695336fbb/pyodide-build/pyodide_build/out_of_tree/utils.py#L29-L39), without changing the behavior.